File: pipeline/adapters/astropy_adapter/experiments/pointing.py
import json
import os
import math
import sys
import time

import numpy as np

try:
    from ..common import ang_sep
except ImportError:
    from common import ang_sep


def _configure_astropy():
    from astropy.utils import iers

    iers.conf.auto_download = False


def _time_from_tt_ut1(jd_tt, jd_ut1, location=None):
    from astropy.time import Time

    t = Time(jd_tt, format="jd", scale="tt", location=location)
    t.delta_ut1_utc = (jd_ut1 - t.utc.jd) * 86400.0
    return t


def _location(lon_rad, lat_rad):
    import astropy.units as u
    from astropy.coordinates import EarthLocation

    return EarthLocation.from_geodetic(
        lon=lon_rad * u.rad,
        lat=lat_rad * u.rad,
        height=0 * u.m,
    )


def _equ_to_horizontal(jd_ut1, jd_tt, ra_rad, dec_rad, obs_lon, obs_lat):
    import astropy.units as u
    from astropy.coordinates import AltAz, SkyCoord, TETE

    loc = _location(obs_lon, obs_lat)
    t = _time_from_tt_ut1(jd_tt, jd_ut1, loc)
    coord = SkyCoord(
        ra=ra_rad * u.rad,
        dec=dec_rad * u.rad,
        frame=TETE(obstime=t, location=loc),
    )
    altaz = coord.transform_to(AltAz(obstime=t, location=loc, pressure=0 * u.hPa))
    back = altaz.transform_to(TETE(obstime=t, location=loc))
    return float(altaz.az.rad), float(altaz.alt.rad), float(back.ra.rad), float(back.dec.rad)


def _horizontal_to_equ(jd_ut1, jd_tt, az_rad, alt_rad, obs_lon, obs_lat):
    import astropy.units as u
    from astropy.coordinates import AltAz, SkyCoord, TETE

    loc = _location(obs_lon, obs_lat)
    t = _time_from_tt_ut1(jd_tt, jd_ut1, loc)
    coord = SkyCoord(
        az=az_rad * u.rad,
        alt=alt_rad * u.rad,
        frame=AltAz(obstime=t, location=loc, pressure=0 * u.hPa),
    )
    equ = coord.transform_to(TETE(obstime=t, location=loc))
    back = equ.transform_to(AltAz(obstime=t, location=loc, pressure=0 * u.hPa))
    equ_back = back.transform_to(TETE(obstime=t, location=loc))
    return (
        float(equ.ra.rad),
        float(equ.dec.rad),
        float(equ_back.ra.rad),
        float(equ_back.dec.rad),
    )


def run_equ_horizontal(lines_iter):
    """Equatorial -> horizontal using Astropy public TETE/AltAz transforms."""
    _configure_astropy()
    n = int(next(lines_iter).strip())
    cases = []

    for _ in range(n):
        parts = next(lines_iter).strip().split()
        jd_ut1 = float(parts[0])
        jd_tt = float(parts[1])
        ra_rad = float(parts[2])
        dec_rad = float(parts[3])
        obs_lon = float(parts[4])
        obs_lat = float(parts[5])

        az, alt, ra_back, dec_back = _equ_to_horizontal(
            jd_ut1, jd_tt, ra_rad, dec_rad, obs_lon, obs_lat
        )

        v_in = np.array([
            math.cos(dec_rad) * math.cos(ra_rad),
            math.cos(dec_rad) * math.sin(ra_rad),
            math.sin(dec_rad),
        ])
        v_bk = np.array([
            math.cos(dec_back) * math.cos(ra_back),
            math.cos(dec_back) * math.sin(ra_back),
            math.sin(dec_back),
        ])
        closure_rad = ang_sep(v_in, v_bk)

        cases.append({
            "jd_ut1": jd_ut1,
            "jd_tt": jd_tt,
            "ra_rad": ra_rad,
            "dec_rad": dec_rad,
            "obs_lon_rad": obs_lon,
            "obs_lat_rad": obs_lat,
            "az_rad": az,
            "alt_rad": alt,
            "closure_rad": closure_rad,
        })

    result = {
        "experiment": "equ_horizontal",
        "library": "astropy",
        "model": "Astropy SkyCoord TETE->AltAz transform",
        "count": n,
        "cases": cases,
    }
    json.dump(result, sys.stdout, indent=None)
    print()


def run_equ_horizontal_perf(lines_iter):
    _configure_astropy()
    n = int(next(lines_iter).strip())
    params = []
    for _ in range(n):
        parts = next(lines_iter).strip().split()
        params.append(tuple(float(p) for p in parts))

    warmup = int(os.environ.get("LAB_PERF_WARMUP", "100"))
    for i in range(min(n, warmup)):
        _equ_to_horizontal(*params[i])

    t0 = time.perf_counter_ns()
    sink = 0.0
    for params_i in params:
        az, alt, _ra_back, _dec_back = _equ_to_horizontal(*params_i)
        sink += az + alt
    elapsed_ns = time.perf_counter_ns() - t0

    result = {
        "experiment": "equ_horizontal_perf",
        "library": "astropy",
        "count": n,
        "total_ns": elapsed_ns,
        "per_op_ns": elapsed_ns / n,
        "throughput_ops_s": n / (elapsed_ns * 1e-9),
        "_sink": float(sink),
    }
    json.dump(result, sys.stdout, indent=None)
    print()


def run_horiz_to_equ(lines_iter):
    """Horizontal -> equatorial using Astropy public AltAz/TETE transforms."""
    _configure_astropy()
    n = int(next(lines_iter).strip())
    cases = []

    for _ in range(n):
        parts = next(lines_iter).strip().split()
        jd_ut1 = float(parts[0])
        jd_tt = float(parts[1])
        az_rad = float(parts[2])
        alt_rad = float(parts[3])
        obs_lon = float(parts[4])
        obs_lat = float(parts[5])

        ra, dec, ra_back, dec_back = _horizontal_to_equ(
            jd_ut1, jd_tt, az_rad, alt_rad, obs_lon, obs_lat
        )

        v_in = np.array([
            math.cos(dec) * math.cos(ra),
            math.cos(dec) * math.sin(ra),
            math.sin(dec),
        ])
        v_bk = np.array([
            math.cos(dec_back) * math.cos(ra_back),
            math.cos(dec_back) * math.sin(ra_back),
            math.sin(dec_back),
        ])
        closure_rad = ang_sep(v_in, v_bk)

        cases.append({
            "jd_ut1": jd_ut1,
            "jd_tt": jd_tt,
            "az_rad": az_rad,
            "alt_rad": alt_rad,
            "obs_lon_rad": obs_lon,
            "obs_lat_rad": obs_lat,
            "ra_rad": ra,
            "dec_rad": dec,
            "closure_rad": closure_rad,
        })

    result = {
        "experiment": "horiz_to_equ",
        "library": "astropy",
        "model": "Astropy SkyCoord AltAz->TETE transform",
        "count": n,
        "cases": cases,
    }
    json.dump(result, sys.stdout, indent=None)
    print()


def run_horiz_to_equ_perf(lines_iter):
    _configure_astropy()
    n = int(next(lines_iter).strip())
    params = []
    for _ in range(n):
        parts = next(lines_iter).strip().split()
        params.append(tuple(float(p) for p in parts))

    warmup = int(os.environ.get("LAB_PERF_WARMUP", "100"))
    for i in range(min(n, warmup)):
        _horizontal_to_equ(*params[i])

    t0 = time.perf_counter_ns()
    sink = 0.0
    for params_i in params:
        ra, dec, _ra_back, _dec_back = _horizontal_to_equ(*params_i)
        sink += ra + dec
    elapsed_ns = time.perf_counter_ns() - t0

    result = {
        "experiment": "horiz_to_equ_perf",
        "library": "astropy",
        "count": n,
        "total_ns": elapsed_ns,
        "per_op_ns": elapsed_ns / n,
        "throughput_ops_s": n / (elapsed_ns * 1e-9),
        "_sink": float(sink),
    }
    json.dump(result, sys.stdout, indent=None)
    print()

File: pipeline/adapters/erfa_adapter/src/pointing.c
#include "common.h"

/* ------------------------------------------------------------------ */
/* Experiment: equ_horizontal                                          */
/* Equatorial (RA/Dec) → Horizontal (Az/Alt) via GAST + eraHd2ae       */
/* Input per line: jd_ut1 jd_tt ra_rad dec_rad obs_lon_rad obs_lat_rad */
/* ------------------------------------------------------------------ */

void run_equ_horizontal(void) {
    int n;
    if (scanf("%d", &n) != 1) { fprintf(stderr, "bad N\n"); exit(1); }

    printf("{\"experiment\":\"equ_horizontal\",\"library\":\"erfa\",");
    printf("\"model\":\"eraHd2ae_GAST\",");
    printf("\"count\":%d,\"cases\":[\n", n);

    for (int i = 0; i < n; i++) {
        double jd_ut1, jd_tt, ra_rad, dec_rad, obs_lon, obs_lat;
        if (scanf("%lf %lf %lf %lf %lf %lf", &jd_ut1, &jd_tt, &ra_rad, &dec_rad,
                  &obs_lon, &obs_lat) != 6) {
            fprintf(stderr, "bad input line %d\n", i); exit(1);
        }
        double ut1_hi = 2451545.0, ut1_lo = jd_ut1 - 2451545.0;
        double tt_hi  = 2451545.0, tt_lo  = jd_tt  - 2451545.0;

        double gast = eraGst06a(ut1_hi, ut1_lo, tt_hi, tt_lo);
        double last = gast + obs_lon;
        double ha   = last - ra_rad;

        double az, alt;
        eraHd2ae(ha, dec_rad, obs_lat, &az, &alt);

        /* Closure */
        double ha_back, dec_back;
        eraAe2hd(az, alt, obs_lat, &ha_back, &dec_back);
        double ra_back = last - ha_back;
        double v_in[3]  = { cos(dec_rad)*cos(ra_rad), cos(dec_rad)*sin(ra_rad), sin(dec_rad) };
        double v_bk[3]  = { cos(dec_back)*cos(ra_back), cos(dec_back)*sin(ra_back), sin(dec_back) };
        double closure_rad = ang_sep(v_in, v_bk);

        if (i > 0) printf(",\n");
        printf("{\"jd_ut1\":%.15f,\"jd_tt\":%.15f,", jd_ut1, jd_tt);
        printf("\"ra_rad\":%.17e,\"dec_rad\":%.17e,", ra_rad, dec_rad);
        printf("\"obs_lon_rad\":%.17e,\"obs_lat_rad\":%.17e,", obs_lon, obs_lat);
        printf("\"az_rad\":%.17e,\"alt_rad\":%.17e,", az, alt);
        printf("\"closure_rad\":%.17e}", closure_rad);
    }
    printf("\n]}\n");
}
void run_equ_horizontal_perf(void) {
    int n;
    if (scanf("%d", &n) != 1) { fprintf(stderr, "bad N\n"); exit(1); }

    double *params = malloc(n * 6 * sizeof(double));

    for (int i = 0; i < n; i++) {
        if (scanf("%lf %lf %lf %lf %lf %lf",
                  &params[6*i], &params[6*i+1], &params[6*i+2],
                  &params[6*i+3], &params[6*i+4], &params[6*i+5]) != 6) {
            fprintf(stderr, "bad input line %d\n", i);
            exit(1);
        }
    }

    /* Warm-up */
    for (int i = 0, warmup = get_perf_warmup(); i < n && i < warmup; i++) {
        double jd_ut1 = params[6*i];
        double jd_tt = params[6*i+1];
        double ra = params[6*i+2];
        double dec = params[6*i+3];
        double lon = params[6*i+4];
        double lat = params[6*i+5];

        double gast = eraGst06a(2451545.0, jd_ut1 - 2451545.0,
                                2451545.0, jd_tt - 2451545.0);
        double ha = normalize_angle(gast + lon - ra);
        double az, alt;
        eraHd2ae(ha, dec, lat, &az, &alt);
    }

    /* Timed run */
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    double sink = 0.0;
    for (int i = 0; i < n; i++) {
        double jd_ut1 = params[6*i];
        double jd_tt = params[6*i+1];
        double ra = params[6*i+2];
        double dec = params[6*i+3];
        double lon = params[6*i+4];
        double lat = params[6*i+5];

        double gast = eraGst06a(2451545.0, jd_ut1 - 2451545.0,
                                2451545.0, jd_tt - 2451545.0);
        double ha = normalize_angle(gast + lon - ra);
        double az, alt;
        eraHd2ae(ha, dec, lat, &az, &alt);
        sink += az;
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed_ns = (t1.tv_sec - t0.tv_sec) * 1e9 + (t1.tv_nsec - t0.tv_nsec);
    double per_op_ns = elapsed_ns / n;

    printf("{\"experiment\":\"equ_horizontal_perf\",\"library\":\"erfa\",");
    printf("\"count\":%d,\"total_ns\":%.0f,\"per_op_ns\":%.1f,", n, elapsed_ns, per_op_ns);
    printf("\"throughput_ops_s\":%.0f,\"_sink\":%.17e}\n",
           (double)n / (elapsed_ns * 1e-9), sink);

    free(params);
}
/* ------------------------------------------------------------------ */
/* Experiment: horiz_to_equ                                            */
/* Horizontal (Az/Alt) → Equatorial (RA/Dec) via eraAe2hd + GAST      */
/* Input: jd_ut1 jd_tt az_rad alt_rad obs_lon_rad obs_lat_rad         */
/* ------------------------------------------------------------------ */

void run_horiz_to_equ(void) {
    int n;
    if (scanf("%d", &n) != 1) { fprintf(stderr, "bad N\n"); exit(1); }

    printf("{\"experiment\":\"horiz_to_equ\",\"library\":\"erfa\",");
    printf("\"model\":\"eraAe2hd_GAST\",");
    printf("\"count\":%d,\"cases\":[\n", n);

    for (int i = 0; i < n; i++) {
        double jd_ut1, jd_tt, az_rad, alt_rad, obs_lon, obs_lat;
        if (scanf("%lf %lf %lf %lf %lf %lf", &jd_ut1, &jd_tt, &az_rad, &alt_rad,
                  &obs_lon, &obs_lat) != 6) {
            fprintf(stderr, "bad input line %d\n", i); exit(1);
        }
        double ut1_hi = 2451545.0, ut1_lo = jd_ut1 - 2451545.0;
        double tt_hi  = 2451545.0, tt_lo  = jd_tt  - 2451545.0;

        double gast = eraGst06a(ut1_hi, ut1_lo, tt_hi, tt_lo);
        double last = gast + obs_lon;

        /* Az/Alt → HA/Dec */
        double ha, dec;
        eraAe2hd(az_rad, alt_rad, obs_lat, &ha, &dec);
        double ra = normalize_angle(last - ha);

        /* Closure: RA/Dec → Az/Alt → RA/Dec */
        double ha2 = last - ra;
        double az2, alt2;
        eraHd2ae(ha2, dec, obs_lat, &az2, &alt2);
        double ha_back, dec_back;
        eraAe2hd(az2, alt2, obs_lat, &ha_back, &dec_back);
        double ra_back = normalize_angle(last - ha_back);

        double v_in[3]  = { cos(dec)*cos(ra), cos(dec)*sin(ra), sin(dec) };
        double v_bk[3]  = { cos(dec_back)*cos(ra_back), cos(dec_back)*sin(ra_back), sin(dec_back) };
        double closure_rad = ang_sep(v_in, v_bk);

        if (i > 0) printf(",\n");
        printf("{\"jd_ut1\":%.15f,\"jd_tt\":%.15f,", jd_ut1, jd_tt);
        printf("\"az_rad\":%.17e,\"alt_rad\":%.17e,", az_rad, alt_rad);
        printf("\"obs_lon_rad\":%.17e,\"obs_lat_rad\":%.17e,", obs_lon, obs_lat);
        printf("\"ra_rad\":%.17e,\"dec_rad\":%.17e,", ra, dec);
        printf("\"closure_rad\":%.17e}", closure_rad);
    }
    printf("\n]}\n");
}

void run_horiz_to_equ_perf(void) {
    int n;
    if (scanf("%d", &n) != 1) { fprintf(stderr, "bad N\n"); exit(1); }

    double *params = malloc(n * 6 * sizeof(double));
    for (int i = 0; i < n; i++) {
        if (scanf("%lf %lf %lf %lf %lf %lf",
                  &params[6*i], &params[6*i+1], &params[6*i+2],
                  &params[6*i+3], &params[6*i+4], &params[6*i+5]) != 6) {
            fprintf(stderr, "bad input line %d\n", i); exit(1);
        }
    }

    for (int i = 0, warmup = get_perf_warmup(); i < n && i < warmup; i++) {
        double jd_ut1 = params[6*i], jd_tt = params[6*i+1];
        double az = params[6*i+2], alt = params[6*i+3];
        double lon = params[6*i+4], lat = params[6*i+5];
        double gast = eraGst06a(2451545.0, jd_ut1 - 2451545.0, 2451545.0, jd_tt - 2451545.0);
        double ha, dec;
        eraAe2hd(az, alt, lat, &ha, &dec);
        (void)ha; (void)dec; (void)gast; (void)lon;
    }

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);
    double sink = 0.0;
    for (int i = 0; i < n; i++) {
        double jd_ut1 = params[6*i], jd_tt = params[6*i+1];
        double az = params[6*i+2], alt = params[6*i+3];
        double lon = params[6*i+4], lat = params[6*i+5];
        double gast = eraGst06a(2451545.0, jd_ut1 - 2451545.0, 2451545.0, jd_tt - 2451545.0);
        double last = gast + lon;
        double ha, dec;
        eraAe2hd(az, alt, lat, &ha, &dec);
        double ra = normalize_angle(last - ha);
        sink += ra + dec;
    }
    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed_ns = (t1.tv_sec - t0.tv_sec) * 1e9 + (t1.tv_nsec - t0.tv_nsec);

    printf("{\"experiment\":\"horiz_to_equ_perf\",\"library\":\"erfa\",");
    printf("\"count\":%d,\"total_ns\":%.0f,\"per_op_ns\":%.1f,", n, elapsed_ns, elapsed_ns / n);
    printf("\"throughput_ops_s\":%.0f,\"_sink\":%.17e}\n",
           (double)n / (elapsed_ns * 1e-9), sink);
    free(params);
}
